concordium-consensus/tests/consensus/ConcordiumTests/LeaderElectionTest.hs

```diff
@@ -0,0 +1,147 @@
+    let nextEpoch = epochSlot + fromIntegral epochLength
+    let toThirds = 2 * fromIntegral epochLength `div` 3
+    -- Find a slot in the last 2/3 of the epoch
+    seedStateSlot :: Slot <- fromIntegral <$> choose (epochSlot + toThirds, nextEpoch-1)
+    seedStateSlotTooEarly :: Slot <- fromIntegral <$> choose (epochSlot, epochSlot + toThirds-1)
+    parentInThisEpoch :: Slot <- fromIntegral <$> choose (fromIntegral seedStateSlot, nextEpoch-1)
+    parentInNextEpoch :: Slot <- fromIntegral <$> choose (nextEpoch, nextEpoch + fromIntegral epochLength-1)
+    slotToPredictFor :: Slot <- fromIntegral <$> choose (fromIntegral parentInNextEpoch, nextEpoch + fromIntegral epochLength-1)
+    slotToPredictTooLate :: Slot <- fromIntegral <$> choose (nextEpoch + fromIntegral epochLength, nextEpoch + 2*fromIntegral epochLength - 1)
+    return $ conjoin [
+                        counterexample "Seedstate slot too early, parent unknown" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly Nothing slotToPredictFor === Nothing,
+                        counterexample "Seedstate slot too early, parent in seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly (Just parentInThisEpoch) slotToPredictFor === Nothing,
+                        counterexample "Seedstate slot too early, parent in epoch after seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly (Just parentInNextEpoch) slotToPredictFor === Nothing,
+                        counterexample "Prediction slot too late, parent unknown" $ predictLeadershipElectionNonce ss seedStateSlot Nothing slotToPredictTooLate === Nothing,
+                        counterexample "Prediction slot too late, parent in seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInNextEpoch) slotToPredictTooLate === Nothing,
+                        counterexample "Prediction slot too late, parent in epoch after seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlot (Just parentInThisEpoch) slotToPredictTooLate === Nothing,
+                        counterexample "Seedstate slot too early and prediction slot too late, parent unknown" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly Nothing slotToPredictTooLate === Nothing,
+                        counterexample "Seedstate slot too early and prediction slot too late, parent in seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly (Just parentInThisEpoch) slotToPredictTooLate === Nothing,
+                        counterexample "Seedstate slot too early and prediction slot too late, parent in epoch after seedstate epoch" $ predictLeadershipElectionNonce ss seedStateSlotTooEarly (Just parentInNextEpoch) slotToPredictTooLate === Nothing
+                    ]
+
+testPredictionJust :: Property
+testPredictionJust = label "Prediction corrrect" makePropertyJust
+
+
+testPredictionNothing :: Property
+testPredictionNothing = label "No prediction possible" makePropertyNothing
```

Comment:
I am not sure that the `label`s add anything here. They're supposed to be for reporting test case distribution, but since there's no non-determinism in the choice, these will always just report 100% for whichever label.